File: Tree/Medium/110-Balanced-Binary-Tree/solution_ResultType.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public class ResultType {
        public bool isBalanced;
        public int depth;
        
        public ResultType(int depth, bool balanced) {
            this.depth = depth;
            this.isBalanced = balanced;
        }
    }
    
    public bool IsBalanced(TreeNode root) {
        // divide and conquer
        // tc:O(n); sc:O(n)
        if(root == null) {
            return true;
        }
        return GetHeight(root).isBalanced;
    }
    
    // return -1 means not balanced binary tree
    private ResultType GetHeight(TreeNode node) {
        if(node == null) {
            return new ResultType(0, true);
        }
        var left = GetHeight(node.left);
        var right = GetHeight(node.right);
        if(!left.isBalanced || !right.isBalanced || Math.Abs(left.depth - right.depth) > 1) {
            return new ResultType(-1, false);
        }
        return new ResultType(Math.Max(left.depth, right.depth) + 1, true);
    }
}

File: Tree/Medium/Lnt597-Subtree-with-Maximum-Average/solution_ResultType.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class ResultType {
public:
    int curSum, numNodes;
    ResultType(int _sum, int _size): curSum(_sum), numNodes(_size) {}
};

class Solution {
public:
    /**
     * @param root: the root of binary tree
     * @return: the root of the maximum average of subtree
     */
    TreeNode * findSubtree2(TreeNode * root) {
        // divide and conquer + Result Type
        // tc:O(n); sc:O(h)
        if(!root) {
            return root;
        }
        TreeNode* maxNode = nullptr;
        double maxAvg = INT_MIN;
        FindMaxTreeNode(root, maxAvg, maxNode);
        return maxNode;
    }

private:
    ResultType FindMaxTreeNode(TreeNode* node, double& maxAvg, TreeNode*& maxNode) {
        if(!node) {
            return ResultType(0, 0);
        }
        ResultType left = FindMaxTreeNode(node->left, maxAvg, maxNode);
        ResultType right = FindMaxTreeNode(node->right, maxAvg, maxNode);
        int sum = node->val + left.curSum + right.curSum;
        int num = left.numNodes + right.numNodes + 1;
        double avg = (double)sum / (double)num;
        if(avg > maxAvg) {
            maxAvg = avg;
            maxNode = node;
        }
        return ResultType(sum, num);
    }
};